ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpCallValidator.kt

```diff
@@ -0,0 +1,108 @@
+package io.ktor.client.features
+
+import io.ktor.client.*
+import io.ktor.client.request.*
+import io.ktor.client.response.*
+import io.ktor.util.*
+import io.ktor.util.pipeline.*
+
+/**
+ * Response validator method.
+ *
+ * You could throw an exception to fail the response.
+ */
+typealias ResponseValidator = suspend (response: HttpResponse) -> Unit
+
+/**
+ * Response exception handler method.
+ */
+typealias CallExceptionHandler = suspend (cause: Throwable) -> Unit
+
+/**
+ * Response validator feature is used for validate response and handle response exceptions.
+ *
+ * See also [Config] for additional details.
+ */
+class HttpCallValidator(
+    private val responseValidators: List<ResponseValidator>,
+    private val callExceptionHandlers: List<CallExceptionHandler>
+) {
+    private suspend fun validateResponse(response: HttpResponse) {
+        responseValidators.forEach { it(response) }
+    }
+
+    private suspend fun processException(cause: Throwable) {
+        callExceptionHandlers.forEach { it(cause) }
+    }
+
+    /**
+     * [HttpCallValidator] configuration.
+     */
+    class Config {
+        internal val responseValidators: MutableList<ResponseValidator> = mutableListOf()
+        internal val responseExceptionHandlers: MutableList<CallExceptionHandler> = mutableListOf()
+
+        /**
+         * Add [CallExceptionHandler].
+         * Last added handler executes first.
+         */
+        fun handleResponseException(block: CallExceptionHandler) {
+            responseExceptionHandlers += block
+        }
+
+        /**
+         * Add [ResponseValidator].
+         * Last added validator executes first.
+         */
+        fun validateResponse(block: ResponseValidator) {
+            responseValidators += block
+        }
+    }
+
+    companion object : HttpClientFeature<Config, HttpCallValidator> {
+        override val key: AttributeKey<HttpCallValidator> = AttributeKey("HttpResponseValidator")
+
+        override fun prepare(block: Config.() -> Unit): HttpCallValidator {
+            val config = Config().apply(block)
+
+            config.responseValidators.reverse()
```

Comment:
Shouldn't we do `config.responseValidators.reversed()` instead to get a reversed defensive copy instead?